src/cmap/auth/gssapi.ts

```diff
@@ -1,82 +1,108 @@
-import { AuthProvider } from './auth_provider';
-import { Kerberos, kModuleError } from '../../deps';
+import { AuthProvider, AuthContext } from './auth_provider';
+import type { Callback } from '../../types';
+import { MongoError } from '../../error';
 
-class GSSAPI extends AuthProvider {
-  auth(authContext: any, callback: Function) {
-    if (Kerberos[kModuleError]) {
-      callback(Kerberos[kModuleError]);
-      return;
-    }
+interface MongoAuthProcessConstructor {
+  new (host: string, port: number, serviceName: string, options: unknown): MongoAuthProcessLike;
+}
+
+interface MongoAuthProcessLike {
+  host: string;
+  port: number;
+  serviceName: string;
+  canonicalizeHostName: boolean;
+  retries: number;
 
-    const { connection, credentials } = authContext;
-    const username = credentials.username;
-    const password = credentials.password;
-    const mechanismProperties = credentials.mechanismProperties;
-    const gssapiServiceName =
-      mechanismProperties['gssapiservicename'] ||
-      mechanismProperties['gssapiServiceName'] ||
-      'mongodb';
+  init: (username: string, password: string, callback: Callback) => void;
+  transition: (payload: unknown, callback: Callback) => void;
+}
 
-    const MongoAuthProcess = Kerberos.processes.MongoAuthProcess;
-    const authProcess = new MongoAuthProcess(
-      connection.host,
-      connection.port,
-      gssapiServiceName,
-      mechanismProperties
-    );
+export class GSSAPI extends AuthProvider {
+  auth(authContext: AuthContext, callback: Callback): void {
+    const { host, port } = authContext.options;
+    if (!host || !port) {
+      return callback(new MongoError('Connection must specify host and port.'));
+    }
 
-    authProcess.init(username, password, (err: any) => {
-      if (err) return callback(err, false);
-      authProcess.transition('', (err?: any, payload?: any) => {
-        if (err) return callback(err, false);
+    import('kerberos')
```

Comment:
should we add this to the `prepare` step instead? I'm not sure this pattern is best for us, but don't feel very strongly about it:
 - its importing the module each time you authenticate
 - it uses promises inside the library, which we generally try not to do
 - nit: it adds an extra level of indentation to this very long method, which is distracting for readability

Is the old pattern in `deps` undesirable for any particular technical reason?

Reply:
> should we add this to the prepare step instead?
Which step
> Is the old pattern in deps undesirable for any particular technical reason?
It's just funky to type as you saw with BSON, I suppose also I was bending over backwards to avoid the type assertion `!` that's not a good reason to start using promises here I suppose 😅 
The module and code would be cached in v8 but I understand why we wouldn't want to lean on that.

> it adds an extra level of indentation to this very long method, which is distracting for readability
The irony of a promise based approach causing this.... 😆 

I'll revert

Reply:
yeah :) this is a tough one! As suggested below, I think we should tackle this as its own piece of work so I created [NODE-2740](https://jira.mongodb.org/browse/NODE-2740)